# รายชื่อ บลจ.

Data : https://api-portal.sec.or.th/apidesc

In [3]:

import http.client
import requests
import urllib.parse
import urllib.error
import base64
import json
import pandas as pd

headers = {
    # Request headers
    'Ocp-Apim-Subscription-Key': "b3238c81040d4850ab6ec07bbd85b8be",
}

params = urllib.parse.urlencode({
})

try:
    conn = http.client.HTTPSConnection('api.sec.or.th')
    response = conn.request("GET", "/FundDailyInfo/amc?%s" % params, "{body}", headers)
    company = pd.read_json(conn.getresponse())     
    conn.close()
except Exception as e:
    print("[Errno {0}] {1}".format(e.errno, e.strerror))


In [2]:
company.head()

last_upd_date                                      name_en  \
0  2019-05-15T05:12:34    KASIKORN ASSET MANAGEMENT COMPANY LIMITED   
1  2019-05-15T05:12:34  MFC ASSET MANAGEMENT PUBLIC COMPANY LIMITED   
2  2019-05-15T05:12:34   MERCHANT PARTNERS ASSET MANAGEMENT LIMITED   
3  2019-05-15T05:12:34         TMB ASSET MANAGEMENT COMPANY LIMITED   
4  2019-05-15T05:12:34         SCB ASSET MANAGEMENT COMPANY LIMITED   

                                             name_th    unique_id  
0        บริษัท หลักทรัพย์จัดการกองทุนกสิกรไทย จำกัด  C0000000021  
1  บริษัท หลักทรัพย์จัดการกองทุนเอ็มเอฟซี จำกัด (...  C0000000023  
2  บริษัท หลักทรัพย์จัดการกองทุนเมอร์ชั่น พาร์ทเน...  C0000000124  
3         บริษัท หลักทรัพย์จัดการกองทุนทหารไทย จำกัด  C0000000182  
4      บริษัท หลักทรัพย์จัดการกองทุนไทยพาณิชย์ จำกัด  C0000000239

In [3]:
company.describe()

last_upd_date                                          name_en  \
count                    26                                               26   
unique                    1                                               26   
top     2019-05-15T05:12:34  UOB ASSET MANAGEMENT (THAILAND) COMPANY LIMITED   
freq                     26                                                1   

                                        name_th    unique_id  
count                                        26           26  
unique                                       26           26  
top     บริษัท หลักทรัพย์จัดการกองทุนภัทร จำกัด  C0000000021  
freq                                          1            1

In [15]:
# เขียนข้อมูลลง Excel

company.to_excel('company.xlsx', index=False)
print('Save Success')

Save Success


In [16]:
comp = pd.read_excel('company.xlsx', 'Sheet1', na_values=['NA'])
comp.head()

last_upd_date                                      name_en  \
0  2019-05-15T05:12:34    KASIKORN ASSET MANAGEMENT COMPANY LIMITED   
1  2019-05-15T05:12:34  MFC ASSET MANAGEMENT PUBLIC COMPANY LIMITED   
2  2019-05-15T05:12:34   MERCHANT PARTNERS ASSET MANAGEMENT LIMITED   
3  2019-05-15T05:12:34         TMB ASSET MANAGEMENT COMPANY LIMITED   
4  2019-05-15T05:12:34         SCB ASSET MANAGEMENT COMPANY LIMITED   

                                             name_th    unique_id  
0        บริษัท หลักทรัพย์จัดการกองทุนกสิกรไทย จำกัด  C0000000021  
1  บริษัท หลักทรัพย์จัดการกองทุนเอ็มเอฟซี จำกัด (...  C0000000023  
2  บริษัท หลักทรัพย์จัดการกองทุนเมอร์ชั่น พาร์ทเน...  C0000000124  
3         บริษัท หลักทรัพย์จัดการกองทุนทหารไทย จำกัด  C0000000182  
4      บริษัท หลักทรัพย์จัดการกองทุนไทยพาณิชย์ จำกัด  C0000000239

# รายชื่อกองทุน

In [18]:
# Loop จาก unique_id ของ บลจ.
import http.client
import requests
import time
import json
import pandas as pd
import urllib.parse
import urllib.error
import base64

headers = {
    # Request headers
    'Ocp-Apim-Subscription-Key': "9a59e81ff953449e9df599624eed1968",
}

params = urllib.parse.urlencode({
})

all_funds = pd.DataFrame(columns=['proj_id', 'proj_abbr_name','proj_name_en', 'proj_name_th','unique_id'])
for unique_id in comp.unique_id:
    req = requests.get(f'https://api.sec.or.th/FundFactsheet/fund/amc/{unique_id}', headers = headers)
    projects = pd.read_json(req.content)
    all_funds = all_funds.append(projects[['proj_id', 'proj_abbr_name','proj_name_en', 'proj_name_th','unique_id']])
    time.sleep(.4)

print(f'There are in total {len(all_funds.index)} funds')

C0000000021
C0000000023
C0000000124
C0000000182
C0000000239
C0000000290
C0000000324
C0000000329
C0000000460
C0000000569
C0000000623
C0000000709
C0000000714
C0000003575
C0000004884
C0000005022
C0000005025
C0000005531
C0000006099
C0000006122
C0000006185
C0000006466
C0000006646
C0000021747
C0000024008
C0000028866
There are in total 10589 funds


In [19]:
all_funds.sample(10)

proj_id   proj_abbr_name  \
14    M0014_2542              RVF   
104   M0426_2550              SVF   
216   M0213_2542            AYFCF   
469   M0288_2551         KTGF12M2   
246   M0080_2562        SCBFF3ME4   
279   M0605_2546          TMBMRMF   
1339  M0699_2553           KKG2YL   
279   M0131_2545             TFGB   
166   M0542_2552  BT-AUTOFIX 3Y/2   
480   M0694_2551         INGMM3M8   

                                       proj_name_en  \
14                             Richee Ventures Fund   
104                           Structured Value Fund   
216                            AYF Chaimongkol Fund   
469   The Krung Thai Government Sector FIF Fund12M2   
246              SCB Foreign Fixed Income Fund 3ME4   
279                TMB Money Retirement Mutual Fund   
1339        K Korean Government Bond 2 Years L Fund   
279                   The TFAM Government Bond Fund   
166                  BT Auto Fixed Income 3Y/2 Fund   
480                            ING Thai MM 3M8 Fund   

                                          proj_name_th    unique_id  
14                                     ริชชี่ เวนเจอส์  C0000000569  
104                           กองทุนเปิดสตรัคเจอร์แวลู  C0000005022  
216                            กองทุนเปิดอยุธยาชัยมงคล  C0000000709  
469   กองทุนรวมกรุงไทยตราสารภาครัฐต่างประเทศ 12 เดือน2  C0000000460  
246      กองทุนเปิดไทยพาณิชย์ตราสารหนี้ต่างประเทศ 3ME4  C0000000239  
279            กองทุนเปิดทหารไทยธนบดีเพื่อการเลี้ยงชีพ  C0000000182  
1339              กองทุนเปิดเค พันธบัตรเกาหลี 2 ปี แอล  C0000000021  
279                กองทุนเปิดทีแฟมกอฟเวิร์นเมนต์ บอนด์  C0000000021  
166                กองทุนเปิดบีที ออโต ตราสารหนี้ 3Y/2  C0000005531  
480               กองทุนเปิด ไอเอ็นจี ไทย เอ็มเอ็ม 3M8  C0000000623

In [20]:
all_funds.describe()

proj_id proj_abbr_name                         proj_name_en  \
count        10589          10589                                10589   
unique       10589          10477                                10493   
top     M0673_2553        TEQT5M1  TISCO Thai Equity Trigger 5M Fund 1   
freq             1              4                                    4   

                                         proj_name_th    unique_id  
count                                           10589        10589  
unique                                          10473           26  
top     กองทุนเปิด ทิสโก้ ไทย อิควิตี้ ทริกเกอร์ 5M#1  C0000000239  
freq                                                4         2183

In [21]:
# เขียนข้อมูลลง Excel

all_funds.to_excel('all_funds.xlsx', index=False)

In [22]:
#เลือก Column
fund_dom = all_funds.loc[:, ['proj_abbr_name',"proj_id"]]
fund_dom.head()

proj_abbr_name     proj_id
0         KFI3MB  M0000_2553
1       KEFF1YAD  M0001_2558
2        KEFI3MK  M0003_2558
3        KFI6MAJ  M0004_2555
4        KFF3MAR  M0005_2560

In [23]:
# เปลี่ยนชื่อ Column
f= fund_dom.rename(index=str, columns={'proj_abbr_name':'Fund Code'})

In [24]:
f.head()

Fund Code     proj_id
0    KFI3MB  M0000_2553
1  KEFF1YAD  M0001_2558
2   KEFI3MK  M0003_2558
3   KFI6MAJ  M0004_2555
4   KFF3MAR  M0005_2560

In [25]:
# อ่านข้อมูลจาก Excel
fund_cat = pd.read_excel('AIMC_Category.xlsx','Thailand Domiciled', na_values=['NA'])
fund_cat.head()

Fund Legal Name  \
0  1 A.M. Flexible Automatic Redemption   
1              Asset Plus Flexible Plus   
2                  Asset Plus Mixed RMF   
3                              B-active   
4                 Bualuang Flexible RMF   

                                   Local Name - Thai     Fund Code  \
0                   กองทุนเปิดวรรณเฟล็กซิเบิลคืนกำไร        FLEXAR   
1            กองทุนเปิด แอสเซทพลัส เฟล็กซิเบิ้ล พลัส  ASP-FLEXPLUS   
2  กองทุนเปิดแอสเซทพลัสหุ้นผสมตราสารหนี้เพื่อการเ...       ASP-MRF   
3                                กองทุนเปิดบีแอ็คทีฟ      B-ACTIVE   
4     กองทุนเปิดบัวหลวงเฟล็กซิเบิ้ลเพื่อการเลี้ยงชีพ        BFLRMF   

                              Firm Name          AIMC Category  
0              One Asset Management Ltd  Aggressive Allocation  
1  Asset Plus Fund Management Co., Ltd.  Aggressive Allocation  
2  Asset Plus Fund Management Co., Ltd.  Aggressive Allocation  
3        BBL Asset Management Co., Ltd.  Aggressive Allocation  
4        BBL Asset Management Co., Ltd.  Aggressive Allocation

In [26]:
fund_cat.describe()

Fund Legal Name  \
count                           1685   
unique                          1685   
top     TISCO China India Retirement   
freq                               1   

                                        Local Name - Thai       Fund Code  \
count                                                1684            1684   
unique                                               1683            1683   
top     กองทุนเปิด แมนูไลฟ์ เอเชีย แปซิฟิก พร็อพเพอร์ต...  ONE-STOXXASEAN   
freq                                                    2               2   

                             Firm Name  AIMC Category  
count                             1685           1684  
unique                              23             42  
top     SCB Asset Management Co., Ltd.  Miscellaneous  
freq                               254            408

In [27]:
#Merge ข้อมูล 2 Dataframe
fund = pd.merge(fund_cat, f ,how='left')
fund.head()

Fund Legal Name  \
0  1 A.M. Flexible Automatic Redemption   
1              Asset Plus Flexible Plus   
2                  Asset Plus Mixed RMF   
3                              B-active   
4                 Bualuang Flexible RMF   

                                   Local Name - Thai     Fund Code  \
0                   กองทุนเปิดวรรณเฟล็กซิเบิลคืนกำไร        FLEXAR   
1            กองทุนเปิด แอสเซทพลัส เฟล็กซิเบิ้ล พลัส  ASP-FLEXPLUS   
2  กองทุนเปิดแอสเซทพลัสหุ้นผสมตราสารหนี้เพื่อการเ...       ASP-MRF   
3                                กองทุนเปิดบีแอ็คทีฟ      B-ACTIVE   
4     กองทุนเปิดบัวหลวงเฟล็กซิเบิ้ลเพื่อการเลี้ยงชีพ        BFLRMF   

                              Firm Name          AIMC Category     proj_id  
0              One Asset Management Ltd  Aggressive Allocation  M0134_2546  
1  Asset Plus Fund Management Co., Ltd.  Aggressive Allocation  M0347_2560  
2  Asset Plus Fund Management Co., Ltd.  Aggressive Allocation  M0785_2547  
3        BBL Asset Management Co., Ltd.  Aggressive Allocation  M0023_2547  
4        BBL Asset Management Co., Ltd.  Aggressive Allocation  M0059_2545

In [28]:
fund.describe()

Fund Legal Name  \
count                             1707   
unique                            1685   
top     TISCO Thai Equity Trigger 5M 1   
freq                                 4   

                                    Local Name - Thai Fund Code  \
count                                            1706      1706   
unique                                           1683      1683   
top     กองทุนเปิด ทิสโก้ ไทย อิควิตี้ ทริกเกอร์ 5M#1   TEQT5M1   
freq                                                4         4   

                             Firm Name  AIMC Category     proj_id  
count                             1707           1706        1341  
unique                              23             42        1340  
top     SCB Asset Management Co., Ltd.  Miscellaneous  M0853_2557  
freq                               255            415           2

In [29]:
fund.sample(10)

Fund Legal Name  \
870                     Bualuang China Equity   
1572                  Asset Plus Fixed Income   
783   Bualuang Global Innovation & Technology   
1636        Multi-National Residence Fund Ord   
1075       Asset Plus Taweesab Fixed Income 9   
1494                    Krung Thai Sri Siri D   
852                   TISCO Global Stars Plus   
338             KTAM SET Commerce ETF Tracker   
1688                      SCB US Equity (I/G)   
902           Bualuang Global Health Care RMF   

                                      Local Name - Thai    Fund Code  \
870                            กองทุนเปิดบัวหลวงหุ้นจีน   B-CHINE-EQ   
1572                   กองทุนเปิดแอสเซท พลัส ตราสารหนี้          ASP   
783      กองทุนเปิดบัวหลวงโกลบอลอินโนเวชั่นและเทคโนโลยี   B-INNOTECH   
1636  กองทุนรวมอสังหาริมทรัพย์มัลติเนชั่นแนลเรสซิเดน...         MNRF   
1075          กองทุนเปิดแอสเซทพลัสตราสารหนี้ทวีทรัพย์ 9  ASP-TFIXED9   
1494         กองทุนเปิดกรุงไทยศรีสิริ ชนิดจ่ายเงินปันผล      KTSRI-D   
852                 กองทุนเปิด ทิสโก้ โกลบอล สตาร์ พลัส       TSTARP   
338            กองทุนเปิด KTAM SET Commerce ETF Tracker        ECOMM   
1688  กองทุนเปิดไทยพาณิชย์หุ้นยูเอส (ชนิดผู้ลงทุนกลุ...   SCBS&P500P   
902   กองทุนเปิดบัวหลวงโกลบอลเฮลธ์แคร์เพื่อการเลี้ยงชีพ     BCARERMF   

                                 Firm Name            AIMC Category  \
870         BBL Asset Management Co., Ltd.     Greater China Equity   
1572  Asset Plus Fund Management Co., Ltd.  Short Term General Bond   
783         BBL Asset Management Co., Ltd.            Global Equity   
1636              MFC Asset Management PLC           Thai Free Hold   
1075  Asset Plus Fund Management Co., Ltd.            Miscellaneous   
1494        Krungthai Asset Management PLC      Moderate Allocation   
852       TISCO Asset Management Co., Ltd.            Global Equity   
338         Krungthai Asset Management PLC           Equity General   
1688        SCB Asset Management Co., Ltd.                US Equity   
902         BBL Asset Management Co., Ltd.              Health Care   

         proj_id  
870   M0651_2560  
1572  M0142_2542  
783   M0667_2559  
1636  M0281_2551  
1075  M0738_2554  
1494         NaN  
852   M0745_2555  
338   M0318_2556  
1688         NaN  
902   M0083_2558

In [30]:
# เขียนข้อมูลลง Excel

fund.to_excel('fund.xlsx', index=False)
print('Save Success')

Save Success


In [55]:
# ตรวจสอบรายละเอียดกองทุนด้วย Code

print(fund[fund['Fund Code']=='TMBPIPF'])

              Fund Legal Name                            Local Name - Thai  \
724  TMB Property Income Plus  กองทุนเปิดทหารไทย พร็อพเพอร์ตี้ อินคัม พลัส   

    Fund Code                     Firm Name  \
724   TMBPIPF  TMB Asset Management Co. Ltd   

                               AIMC Category     proj_id  
724  Fund of Property fund -Thai and Foreign  M0221_2557  


{"last_upd_date":"2019-05-14T05:09:23","nav_date":"2019-05-13","net_asset":16678081140.10,"last_val":104.1768,"previous_val":104

# จัดการข้อมูล NAV ที่ดาวน์โหลดมาจากระบบ

In [58]:
import pandas as pd
import time
import base64

In [59]:
def read_data(month):
    
    n = ['nav_date','B','C','Firm','E','proj_name_th','proj_name_en','fund_code','net_asset',
     'nav_price','dividend_value','book_close_date','sell_price','buy_price']
    df = pd.read_csv((f'download/{month}-NAV.txt'),header=None,names=n)
    
    # Convert Type to DateTime
    df['nav_date'] = pd.to_datetime(df.nav_date,dayfirst=True ) 
    df['book_close_date'] = pd.to_datetime(df.book_close_date,dayfirst=True ) 
    return df

In [60]:
# กำหนดไฟล์ที่ต้องการดึงข้อมูล (ต้อง save file ให้เป็น UTF-8 ก่อน)

all_nav = pd.DataFrame(columns=['nav_date',"fund_code",'nav_price','net_asset','sell_price',
                             'buy_price','dividend_value','book_close_date','Firm','proj_name_th','proj_name_en'])
y=['2019-05','2019-04','2019-03','2019-02','2019-01']

for i in range(len(y)):
    df = read_data(y[i])
    df = df.loc[:, ['nav_date',"fund_code",'nav_price','net_asset','sell_price',
                             'buy_price','dividend_value','book_close_date','Firm','proj_name_th','proj_name_en']]
    all_nav = all_nav.append(df[['nav_date',"fund_code",'nav_price','net_asset','sell_price',
                             'buy_price','dividend_value','book_close_date','Firm','proj_name_th','proj_name_en']])

all_nav.sample(10)    

nav_date               fund_code  nav_price     net_asset  sell_price  \
24577 2019-04-29     CIMB-PRINCIPAL iDIV        NaN  2.001341e+09         NaN   
208   2019-05-02     CIMB-PRINCIPAL TDIF        NaN  7.229969e+08         NaN   
7224  2019-03-08            KT-BRAIN RMF     9.8323  3.506922e+07      9.8324   
11666 2019-04-12  CIMB-PRINCIPAL GOLD8P2     7.7201  1.949578e+07         NaN   
3387  2019-03-05             M-TWSUK RMF    10.0203  1.982761e+07     10.0204   
9882  2019-03-12                    KASF    15.1371  1.701497e+09     15.2507   
11213 2019-01-14               SCBFF1YL4    10.0050  2.716415e+09         NaN   
15528 2019-01-18              B-INNOTECH    11.6655  3.171902e+09     11.7822   
23072 2019-03-25                 TMB50DV     8.7881  1.789168e+09      8.7970   
3361  2019-02-05              SCBAEMFUND    10.8302  3.084301e+09     10.9462   

       buy_price  dividend_value book_close_date            Firm  \
24577        NaN             NaN             NaT  CIMB-PRINCIPAL   
208          NaN             NaN             NaT  CIMB-PRINCIPAL   
7224      9.8323             NaN             NaT            KTAM   
11666     7.7201             NaN             NaT  CIMB-PRINCIPAL   
3387     10.0203             NaN             NaT             MFC   
9882     15.1371             NaN             NaT          KASSET   
11213        NaN             NaN             NaT           SCBAM   
15528    11.6655             NaN             NaT           BBLAM   
23072     8.7793            0.13      2018-01-03           TMBAM   
3361     10.8302             NaN             NaT           SCBAM   

                                            proj_name_th  \
24577  กองทุนเปิดซีไอเอ็มบี-พรินซิเพิล อิควิตี้ ดิวิเ...   
208    กองทุนเปิดซีไอเอ็มบี-พรินซิเพิล ไทย ไดนามิก อิ...   
7224       กองทุนเปิดกรุงไทย เอไอ เบรน เพื่อการเลี้ยงชีพ   
11666  กองทุนเปิดซีไอเอ็มบี-พรินซิเพิล โกลด์ 8 เปอร์เ...   
3387   กองทุนเปิดเอ็มเอฟซี ออมเกษียณทวีสุข เพื่อการเล...   
9882              กองทุนเปิดเคเอ ชอร์ท เทอม ฟิกซ์ อินคัม   
11213      กองทุนเปิดไทยพาณิชย์ตราสารหนี้ต่างประเทศ 1YL4   
15528     กองทุนเปิดบัวหลวงโกลบอลอินโนเวชั่นและเทคโนโลยี   
23072                      กองทุนเปิดทหารไทย SET50 ปันผล   
3361   กองทุนเปิดไทยพาณิชย์ เอเชียน อีเมอร์จิ้ง มาร์เ...   

                                          proj_name_en  
24577       CIMB-Principal Equity Dividend Income Fund  
208    CIMB-Principal Thai Dynamic Income Equity  Fund  
7224                           Krung Thai AI Brain RMF  
11666     CIMB-Principal Gold 8 Percent Trigger 2 Fund  
3387               MFC Senior Taweesuk Retirement Fund  
9882                   KA Short Term Fixed-Income Fund  
11213               SCB Foreign Fixed Income Fund 1YL4  
15528     Bualuang Global Innovation & Technology Fund  
23072                          TMB SET50 Dividend Fund  
3361          SCB ASIAN EMERGING MARKETS OPEN END FUND

In [61]:
all_nav.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 121978 entries, 0 to 28303
Data columns (total 11 columns):
nav_date           121978 non-null datetime64[ns]
fund_code          121978 non-null object
nav_price          120882 non-null float64
net_asset          121974 non-null float64
sell_price         87893 non-null float64
buy_price          88041 non-null float64
dividend_value     19646 non-null float64
book_close_date    19735 non-null datetime64[ns]
Firm               121978 non-null object
proj_name_th       121978 non-null object
proj_name_en       121978 non-null object
dtypes: datetime64[ns](2), float64(5), object(4)
memory usage: 11.2+ MB


In [62]:
# หาจำนวนกองทุน

len(all_nav['fund_code'].unique())

1534

In [63]:
# หาจำนวนวัน

len(all_nav['nav_date'].unique())

91

In [64]:
# นับจำนวน Duplicate

len(all_nav[ all_nav.duplicated(['nav_date', 'fund_code'], keep = False) ])

0

In [65]:
# ลบ Duplicate

# Remove the duplicates
all_nav.drop_duplicates(['nav_date', 'fund_code'], inplace=True)

# Reset dataframe index after drop_duplicates.
all_nav.reset_index(drop=True, inplace=True)
all_nav.index += 1

len(all_nav)

121978

In [66]:
all_nav.head()

nav_date fund_code  nav_price     net_asset  sell_price  buy_price  \
1 2019-05-02      MJLF    12.1474  4.008653e+09         NaN        NaN   
2 2019-05-02    CTARAF    10.1526  3.248834e+09         NaN        NaN   
3 2019-05-02    GOLDPF    11.0148  2.269067e+09         NaN        NaN   
4 2019-05-02     KPNPF    10.2108  1.837954e+09         NaN        NaN   
5 2019-05-02      RGPF    25.3412  2.075008e+10         NaN        NaN   

   dividend_value book_close_date    Firm  \
1          0.2300      2013-11-26  KASSET   
2          0.0430      2013-08-28  KASSET   
3          0.1360      2013-08-28  KASSET   
4          0.1625      2013-11-26  KASSET   
5             NaN             NaT    KTAM   

                                        proj_name_th  \
1  กองทุนรวมสิทธิการเช่าอสังหาริมทรัพย์เมเจอร์ ซี...   
2  กองทุนรวมสิทธิการเช่าอสังหาริมทรัพย์โรงแรมและร...   
3          กองทุนรวมสิทธิการเช่าอสังหาริมทรัพย์โกลด์   
4                   กองทุนรวมอสังหาริมทรัพย์เคพีเอ็น   
5                     รีเจนท์โกลด์พร็อพเพอร์ตี้ฟันด์   

                                       proj_name_en  
1  Major Cineplex Lifestyle Leasehold Property Fund  
2  Centara Hotels & Resorts Leasehold Property Fund  
3                   Gold Property Fund (Lease Hold)  
4                                 KPN Property Fund  
5                         Regent Gold Property Fund

# Create Database

- https://docs.python.org/3/library/sqlite3.html
- http://sebastianraschka.com/Articles/2014_sqlite_in_python_tutorial.html#creating-a-new-sqlite-database
-https://www.dataquest.io/blog/python-pandas-databases/
- SQL Basics: Working with Databases : https://www.dataquest.io/blog/sql-basics/


In [ ]:
import sqlite3, json
from sqlite3 import Error

def create_connection(database):
    try:
        conn = sqlite3.connect(database, isolationlevel=None, checksamethread = False)
        conn.row_factory = lambda c, r: dict(zip(_[_col_[_0] for col in c.description], r))

        return conn
    except Error as e:
        print(e)

def create_table(c):
    sql = """ 
        CREATE TABLE IF NOT EXISTS articles (
            id integer PRIMARY KEY,
            name varchar(225) NOT NULL,
            body varchar(255) NOT NULL
        ); 
    """
    
# ***   'nav_date',"fund_code",'nav_price','net_asset','sell_price',
 #                            'buy_price','dividend_value','book_close_date','Firm','proj_name_th','proj_name_en'
    
    
    c.execute(sql)

def insert_db(c, name, body):
    sql = ''' INSERT INTO articles(name, body)
                VALUES (?, ?) '''
    c.execute(sql, (name, body))

def select_all_items(c, name):
    c.execute("SELECT * FROM articles WHERE name like ?", ('%'+name+'%',))
    rows = c.fetchall()
    return rows

def main():
    database = "./sqlite.db"

    # create a database connection
    conn = create_connection(database)

    # create items table
    create_table(conn)

    # confirm that table is created
    print("Connection established!")

if _**name == '**_main_':
    main()

In [54]:
print(all_nav[all_nav['fund_code']=='TMB50'])

         nav_date fund_code  nav_price     net_asset  sell_price  buy_price  \
1282   2019-02-05     TMB50   106.3812  1.712551e+10    106.4877   106.2748   
2577   2019-03-05     TMB50   106.5281  1.712442e+10    106.6347   106.4216   
4053   2019-07-05     TMB50   105.8674  1.697361e+10    105.9734   105.7615   
5526   2019-08-05     TMB50   104.9219  1.681371e+10    105.0269   104.8170   
6951   2019-09-05     TMB50   104.4093  1.674946e+10    104.5138   104.3049   
8430   2019-10-05     TMB50   104.6432  1.679140e+10    104.7479   104.5386   
9860   2019-05-13     TMB50   104.1768  1.667808e+10    104.2811   104.0726   
11198  2019-05-14     TMB50   103.8788  1.651101e+10    103.9828   103.7749   
12418  2019-05-15     TMB50   103.1196  1.640358e+10    103.2228   103.0165   
13847  2019-01-04     TMB50   103.7923  1.625069e+10    103.8962   103.6885   
15297  2019-02-04     TMB50   104.2680  1.630803e+10    104.3724   104.1637   
16744  2019-03-04     TMB50   104.2297  1.636694e+10

In [14]:
n = ['nav_date','B','C','Firm','E','proj_name_th','proj_name_en','fund_code','net_asset',
     'nav_price','dividend_value','book_close_date','sell_price','buy_price']
get_nav = pd.read_csv('download/2019-05-NAV.txt',header=None,names=n,encoding='utf-8')
get_nav.sample(5)

nav_date   B  C             Firm                E  \
4840   08/05/2019  OE                KSAM             KSAM   
6938   09/05/2019  OE               TMBAM            TMBAM   
3424   07/05/2019  OE          MANULIFEAM       MANULIFEAM   
5318   08/05/2019  OE     THANACHART FUND  THANACHART FUND   
10179  14/05/2019  OE               BBLAM            BBLAM   

                                            proj_name_th  \
4840        กองทุนเปิดกรุงศรีชีวิตดี๊ดีเพื่อการเลี้ยงชีพ   
6938     กองทุนเปิดทหารไทย Emerging Markets Equity Index   
3424   กองทุนเปิด แมนูไลฟ์ เอเชีย แปซิฟิก พร็อพเพอร์ต...   
5318                         กองทุนเปิดธนชาต Income Plus   
10179                   กองทุนเปิดบัวหลวงทศพลหุ้นระยะยาว   

                                   proj_name_en     fund_code     net_asset  \
4840                    Krungsri Happy Life RMF    KFHAPPYRMF  6.607619e+08   
6938     TMB Emerging Markets Equity Index Fund       TMBEMEQ  5.352372e+08   
3424   Manulife Asia Pacific Property REIT Fund     MN-APREIT  4.502194e+08   
5318                Thanachart Income Plus Fund  T-IncomePlus  4.912926e+10   
10179                      Bualuang Top-Ten LTF   B-TOPTENLTF  3.131592e+08   

       nav_price  dividend_value book_close_date  sell_price  buy_price  
4840     10.0890             NaN             NaN     10.0891    10.0890  
6938      9.8086             NaN             NaN      9.8185     9.7988  
3424     12.0030             NaN             NaN     12.1831    12.0030  
5318     11.2442             NaN             NaN     11.2443    11.2442  
10179    10.0051             NaN             NaN     10.0052    10.0051

In [15]:
#เลือก Column
nav = get_nav.loc[:, ['nav_date',"fund_code",'nav_price','net_asset','sell_price',
                             'buy_price','dividend_value','book_close_date','Firm','proj_name_th','proj_name_en']]
nav.head()

nav_date fund_code  nav_price     net_asset  sell_price  buy_price  \
0  02/05/2019      MJLF    12.1474  4.008653e+09         NaN        NaN   
1  02/05/2019    CTARAF    10.1526  3.248834e+09         NaN        NaN   
2  02/05/2019    GOLDPF    11.0148  2.269067e+09         NaN        NaN   
3  02/05/2019     KPNPF    10.2108  1.837954e+09         NaN        NaN   
4  02/05/2019      RGPF    25.3412  2.075008e+10         NaN        NaN   

   dividend_value book_close_date    Firm  \
0          0.2300      26/11/2013  KASSET   
1          0.0430      28/08/2013  KASSET   
2          0.1360      28/08/2013  KASSET   
3          0.1625      26/11/2013  KASSET   
4             NaN             NaN    KTAM   

                                        proj_name_th  \
0  กองทุนรวมสิทธิการเช่าอสังหาริมทรัพย์เมเจอร์ ซี...   
1  กองทุนรวมสิทธิการเช่าอสังหาริมทรัพย์โรงแรมและร...   
2          กองทุนรวมสิทธิการเช่าอสังหาริมทรัพย์โกลด์   
3                   กองทุนรวมอสังหาริมทรัพย์เคพีเอ็น   
4                     รีเจนท์โกลด์พร็อพเพอร์ตี้ฟันด์   

                                       proj_name_en  
0  Major Cineplex Lifestyle Leasehold Property Fund  
1  Centara Hotels & Resorts Leasehold Property Fund  
2                   Gold Property Fund (Lease Hold)  
3                                 KPN Property Fund  
4                         Regent Gold Property Fund

In [17]:
nav.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12567 entries, 0 to 12566
Data columns (total 11 columns):
nav_date           12567 non-null object
fund_code          12567 non-null object
nav_price          12461 non-null float64
net_asset          12567 non-null float64
sell_price         8465 non-null float64
buy_price          8473 non-null float64
dividend_value     1881 non-null float64
book_close_date    1888 non-null object
Firm               12567 non-null object
proj_name_th       12567 non-null object
proj_name_en       12567 non-null object
dtypes: float64(5), object(6)
memory usage: 1.1+ MB


In [6]:
# ตรวจสอบรายละเอียดราคากองทุนด้วย Code

print(nav[nav['fund_code']=='1AMSET50'])

         nav_date fund_code  nav_price     net_asset  sell_price  buy_price  \
832    01/04/2019  1AMSET50    63.2937  4.353950e+09         NaN        NaN   
2279   02/04/2019  1AMSET50    63.6140  4.379573e+09         NaN        NaN   
3726   03/04/2019  1AMSET50    63.5605  4.411098e+09         NaN        NaN   
5174   04/04/2019  1AMSET50    63.2264  4.388184e+09         NaN        NaN   
6593   05/04/2019  1AMSET50    63.3248  4.396287e+09         NaN        NaN   
8042   09/04/2019  1AMSET50    63.7691  4.429888e+09         NaN        NaN   
9495   10/04/2019  1AMSET50    63.9199  4.441431e+09         NaN        NaN   
10855  11/04/2019  1AMSET50    63.7924  4.431180e+09         NaN        NaN   
12323  12/04/2019  1AMSET50    63.8318  4.441805e+09         NaN        NaN   
13785  17/04/2019  1AMSET50    64.3152  4.478603e+09         NaN        NaN   
15242  18/04/2019  1AMSET50    64.4629  4.490095e+09         NaN        NaN   
16594  19/04/2019  1AMSET50    64.4599  4.488756e+09

In [53]:
s= nav.fund_code.unique()
len(s)

1516

In [54]:
nav.describe()

nav_price     net_asset    sell_price     buy_price  dividend_value
count  27073.000000  2.730700e+04  18763.000000  18862.000000     4167.000000
mean      14.816568  3.194239e+09     15.903436     15.888484        0.505225
std       35.997297  1.372921e+10     36.114420     36.045795        0.746531
min        0.943100  9.723982e+05      0.947900      0.943100        0.015500
25%       10.001300  1.684373e+08     10.022000      9.929675        0.178200
50%       10.445500  7.448199e+08     11.234000     11.103600        0.290000
75%       12.667500  2.483801e+09     14.251250     14.122175        0.550000
max     1248.941700  3.826959e+11   1251.439700   1246.443800        8.500000

# ค้นหาชื่อกองทุนด้วยชื่อย่อหรือชื่อกองทุนบางส่วน

In [ ]:
import http.client, urllib.request, urllib.parse, urllib.error, base64

headers = {
    # Request headers
    'Content-Type': 'application/json',
    'Ocp-Apim-Subscription-Key': '9a59e81ff953449e9df599624eed1968',
}

params = urllib.parse.urlencode({
})

try:
    conn = http.client.HTTPSConnection('api.sec.or.th')
    conn.request("POST", "/FundFactsheet/fund?%s" % params, "{name=TMB50}", headers) # ทดสอบ 'TMB50'
    #https://api.sec.or.th/FundFactsheet/fund?proj_abbr_name=TMB50
    response = conn.getresponse()
    data = response.read()
    print(data)
    conn.close()
except Exception as e:
    print("[Errno {0}] {1}".format(e.errno, e.strerror))


In [ ]:
# Data : https://api-portal.sec.or.th/apidesc

import http.client
import requests
import urllib.parse
import urllib.error
import base64
import json
import pandas as pd

headers = {
    # Request headers
    'Ocp-Apim-Subscription-Key': "9a59e81ff953449e9df599624eed1968",
}

params = urllib.parse.urlencode({
})

#for i in range(len(comp)):
try:
    conn = http.client.HTTPSConnection('api.sec.or.th')
    response = conn.request("GET", "/FundFactsheet/fund/amc/C0000000021?%s" % params, "{body}", headers)
    response = conn.getresponse()
    fund = response.read()  
    j = pd.io.json.loads(fund)
    ja = ja
    conn.close()
except Exception as e:
    print("[Errno {0}] {1}".format(e.errno, e.strerror))


In [ ]:
print(fund)

In [ ]:
print(comp)

In [ ]:
import pandas as pd
import json
import requests
%matplotlib inline
import time

In [ ]:
all_funds = []
for i in range(len(comp)):
    url = (f'https://api.sec.or.th/FundFactsheet/fund/amc/{comp[i]}')
    headers = {'Ocp-Apim-Subscription-Key': '9a59e81ff953449e9df599624eed1968'}
    response = requests.request('GET', url,headers=headers)
    projects = pd.read_json(response.content)
    time.sleep(500)
    all_funds = all_funds.append(projects[['proj_id', 'proj_abbr_name','proj_name_en', 'proj_name_th','unique_id']])


In [ ]:
print(all_funds)

In [ ]:
f

In [ ]:
GET https://api.sec.or.th/FundFactsheet/fund/amc/C0000000623 

In [ ]:
f = open('company.txt','w',encoding='utf-8')
f.write (str(data))
#หรือ f.writelines(s)
f.close()